### PR TITLE
Fix StackOverflowError on types with self-references

### PR DESCRIPTION
**Pull request by psteiger**
**Description**:

When we have self types (implemented by recursive generics), Motif fails with `StackOverflowError`:
```
	at com.uber.xprocessing.ext.XTypeKt.hasCollectionType(XType.kt:316)
	at com.uber.xprocessing.ext.XTypeKt.hasCollectionType(XType.kt:316)
	at com.uber.xprocessing.ext.XTypeKt.hasCollectionType(XType.kt:316)
        ...
```
The following code seems to trigger that condition:
```kotlin
open class SchedulePluginRouter<out R : SchedulePluginRouter<R, *>, I : SchedulePluginInteractor<*, *, *>>(
    interactor: I,
) : Router<I>(interactor)

abstract class SchedulePluginInteractor<P : Any, I : SchedulePluginInteractor<*, I, *>, R : SchedulePluginRouter<R, I>>(
    presenter: P,
) : Interactor<P, R>(presenter) 
```

This PR changes `XType.hasCollectionType()` implementation to:
1. Not re-process types that were already processed
2. Use a queue instead of recursion (to avoid StackOverflowError for big enough type trees).

### CODE REVIEW

**Comment by davissuber** [comment]
Code looks great; could we add a test?

File: compiler/ast/src/main/kotlin/com/uber/xprocessing/ext/XType.kt [hunk 1]
```diff
@@ -312,8 +312,16 @@ fun XType.isPrimitive(): Boolean {
 }
 
 private fun XType.hasCollectionType(): Boolean {
-  return this.typeElement?.name.orEmpty() in collectionTypes ||
-      typeArguments.any { it.hasCollectionType() }
+  val visited = mutableSetOf<XType>()
+  val queue = mutableListOf(this@hasCollectionType)
+  while (queue.isNotEmpty()) {
+    val type = queue.removeFirst()
+    if (type in visited) continue
```
Comment (tyvsmith) [inline compiler/ast/src/main/kotlin/com/uber/xprocessing/ext/XType.kt:319]: How big will visited get? This becomes an O(N^2) right? Do we move to something constant since this will effect build times?

File: compiler/ast/src/main/kotlin/com/uber/xprocessing/ext/XType.kt [hunk 2]
```diff
@@ -312,8 +312,16 @@ fun XType.isPrimitive(): Boolean {
 }
 
 private fun XType.hasCollectionType(): Boolean {
-  return this.typeElement?.name.orEmpty() in collectionTypes ||
-      typeArguments.any { it.hasCollectionType() }
+  val visited = mutableSetOf<XType>()
```
Comment (tyvsmith) [inline compiler/ast/src/main/kotlin/com/uber/xprocessing/ext/XType.kt:315]: Was our xprocessing copied out of Room? If so, do we need to keep this up to date with the upstream version and a fix is needed there (or is already fixed and we should pull it)?

File: compiler/ast/src/main/kotlin/com/uber/xprocessing/ext/XType.kt (hunk 1 above)
Comment (psteiger) [reply compiler/ast/src/main/kotlin/com/uber/xprocessing/ext/XType.kt:319]: `visited` uses a `mutableSetOf` which uses `LinkedHashSet` behind the curtains. 
It's O(1) (constant) time complexity both for adding and for checking if set contains some element.

I believe the auxiliary list works the same as the previous recursive solution (difference is instead of leveraging the call stack, I'm using my own list), so no complexity degradation here as well.

So, all in all, I don't expect any build time degradation (I took that into consideration when designing the solution, but didn't bother to collect any metrics/benchmarking)

File: compiler/ast/src/main/kotlin/com/uber/xprocessing/ext/XType.kt (hunk 2 above)
Comment (psteiger) [reply compiler/ast/src/main/kotlin/com/uber/xprocessing/ext/XType.kt:315]: I'm not sure if we copied out of Room. I admit I'm opening a PR here without much context on the big picture, history etc. 

@jbarr21 would you know?

Comment (jbarr21) [reply compiler/ast/src/main/kotlin/com/uber/xprocessing/ext/XType.kt:315]: We consume xprocessing from Google's maven repo, though there are likely newer versions available 
https://github.com/uber/motif/blob/main/gradle/dependencies.gradle#L28

**Review by jbarr21 — APPROVED** [review]

**Comment by davissuber** [comment]
Some extra investigation notes regarding the room-compiler-processing library:
 * the version are on `2.6.0-alpha02` is [closer to the latest version](https://androidx.tech/artifacts/room/room-compiler-processing/) `2.6.1` than I originally thought
 * there aren't that much difference in the XType neighborhood between the two versions (based on 5-min of my untrained eyeballing), so bumping the library alone may not solve the stack overflow problem
 * however, based on [this comment](https://github.com/uber/motif/blob/main/compiler/ast/src/main/kotlin/com/uber/xprocessing/ext/XType.kt#L170), it seems that `hasCollectionType` is introduced in part to handle some KSP type equality checks, which [even the current version of room-compiler-processing has some special override](https://androidx.tech/artifacts/room/room-compiler-processing/2.6.0-alpha02-source/androidx/room/compiler/processing/ksp/KspType.kt.html)
    * in other words, maybe we can circumvent the stack overflow problem by changing how we use the room-compiler-processing library (hand-wavy comments made without much context)
